### PR TITLE
Fix payloadsIndex cleanup

### PR DESCRIPTION
**Pull request by Mr-Leshiy**

### CODE REVIEW

File: src/pop/blockchain/alt_block_tree.cpp [hunk 1]
```diff
@@ -699,10 +699,22 @@ AltBlockTree::AltBlockTree(const AltBlockTree::alt_config_t& alt_config,
...
+  base::removeSubtree(toRemove, &onRemove);
+}
+
 void AltBlockTree::removeSubtree(AltBlockTree::index_t& toRemove) {
   VBK_TRACE_ZONE_SCOPED;
-  payloadsIndex_.removePayloadsIndex(toRemove);
-  base::removeSubtree(toRemove);
+  std::function<void(BlockIndex<AltBlock>&)> onRemove =
+      [&](const BlockIndex<AltBlock>& next) {
+        payloadsIndex_.removePayloadsIndex(next);
```
Comment (blackyblack) [inline src/pop/blockchain/alt_block_tree.cpp:715]: `next` does not look like a proper name. It looks like only next blocks are being affected, skipping `toRemove` block.

File: src/pop/blockchain/pop/vbk_block_tree.cpp [hunk 2]
```diff
@@ -416,10 +416,22 @@ bool VbkBlockTree::loadBlockInner(const stored_index_t& index,
   return true;
 }
 
+void VbkBlockTree::removeSubtree(const VbkBlockTree::hash_t& toRemove) {
+  VBK_TRACE_ZONE_SCOPED;
+  std::function<void(BlockIndex<VbkBlock>&)> onRemove =
+      [&](const BlockIndex<VbkBlock>& next) {
+        payloadsIndex_.removePayloadsIndex(next);
+      };
+  BaseBlockTree::removeSubtree(toRemove, &onRemove);
```
Comment (blackyblack) [inline src/pop/blockchain/pop/vbk_block_tree.cpp:425]: Do not duplicate the code. Find block index by hash and call `removeSubtree(VbkBlockTree::index_t& toRemove)`

File: src/pop/blockchain/pop/vbk_block_tree.cpp [hunk 3]
```diff
@@ -416,10 +416,22 @@ bool VbkBlockTree::loadBlockInner(const stored_index_t& index,
...
+
 void VbkBlockTree::removeSubtree(VbkBlockTree::index_t& toRemove) {
   VBK_TRACE_ZONE_SCOPED;
-  payloadsIndex_.removePayloadsIndex(toRemove);
-  BaseBlockTree::removeSubtree(toRemove);
+  std::function<void(BlockIndex<VbkBlock>&)> onRemove =
+      [&](const BlockIndex<VbkBlock>& next) {
+        payloadsIndex_.removePayloadsIndex(next);
+      };
+  BaseBlockTree::removeSubtree(toRemove, &onRemove);
```
Comment (blackyblack) [inline src/pop/blockchain/pop/vbk_block_tree.cpp:434]: Is `base::removeSubtree` not working here?

**Review by blackyblack — CHANGES_REQUESTED** [review]

File: include/veriblock/pop/blockchain/base_block_tree.hpp [hunk 4]
```diff
@@ -1079,7 +1081,11 @@ struct BaseBlockTree {
     }
   }
 
+  virtual void onSingleBlockRemove(const index_t&) {}
```
Comment (Warchant) [inline include/veriblock/pop/blockchain/base_block_tree.hpp:1084]: Add docs

**Review by Warchant — APPROVED** [review]
Please rename a method to onBeforeLeafRemoved and add docs to base method, and then this is good to be merged!

File: src/pop/blockchain/pop/vbk_block_tree.cpp (hunk 2 above)
Comment (Mr-Leshiy) [reply src/pop/blockchain/pop/vbk_block_tree.cpp]: changed

File: src/pop/blockchain/alt_block_tree.cpp (hunk 1 above)
Comment (Mr-Leshiy) [reply src/pop/blockchain/alt_block_tree.cpp]: changed

File: src/pop/blockchain/pop/vbk_block_tree.cpp (hunk 3 above)
Comment (Mr-Leshiy) [reply src/pop/blockchain/pop/vbk_block_tree.cpp]: changed

**Pull request merged**